File: testdata/precision-suite-elixir/tp_ssrf.exs
# SSRF (CWE-918): an untrusted value is used as the request URL of an outbound
# HTTP client (HTTPoison / :httpc / Req). A correct scanner fires TAINT-006 on
# each sink line.

defmodule TpSsrf do
  def fetch_httpoison(conn) do
    url = conn.params["url"]
    HTTPoison.get(url) # nox-expect: TAINT-006
  end

  def fetch_httpc(conn) do
    target = conn.query_params
    :httpc.request(target) # nox-expect: TAINT-006
  end

  def fetch_req(conn) do
    endpoint = conn.body_params
    Req.get(endpoint) # nox-expect: TAINT-006
  end
end

File: testdata/precision-suite-clojure/tp_ssrf.clj
(ns app.ssrf
  "SSRF (CWE-918): a tainted URL is fetched with clj-http. Unlike command
   injection, validation of the string does not help — the request still goes to
   the attacker-chosen host — so a correct scanner fires TAINT-006."
  (:require [clj-http.client :as client]))

;; clj-http GET of a request-controlled URL.
(defn fetch [req]
  (let [url (:params req)]
    (client/get url))) ; nox-expect: TAINT-006

;; A POST to a tainted URL is equally an SSRF vector.
(defn push [req]
  (let [target (:query-string req)]
    (client/post target))) ; nox-expect: TAINT-006
